### PR TITLE
ROX-12875: Add indexes for SAC filters

### PR DESCRIPTION
**Pull request by janisz**
## Description

To speed up SAC filtering this PR adds hash indexes on Cluster and Namespaces.

## Testing Performed

CI

### CODE REVIEW

**Comment by janisz** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3238** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/3238" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/3238?utm_source=stack-comment).

**Comment by connorgorman** [comment]
Old is sept 19th master run and new is this run. I wonder if just a targeted SAC approach may be better. Like indexes 
 on deployment may be the most beneficial due to the joins generally terminating at deployments
```
sensor_event_duration Action=CREATE_RESOURCE Type=AlertResults (old: (25220.89/2961) 8.52, new (28467.28/2846) 10.00): change: 17.4327%
sensor_event_duration Action=CREATE_RESOURCE Type=Deployment (old: (64606.62/2499) 25.85, new (73272.71/2496) 29.36): change: 13.5499%
sensor_event_duration Action=CREATE_RESOURCE Type=Pod (old: (282955.83/124824) 2.27, new (331740.21/112650) 2.94): change: 29.9111%
sensor_event_duration Action=CREATE_RESOURCE Type=ProcessIndicator (old: (15980.20/485431) 0.03, new (13545.84/427666) 0.03): change: -3.7842%
sensor_event_duration Action=REMOVE_RESOURCE Type=AlertResults (old: (69018.71/5000) 13.80, new (73930.34/5000) 14.79): change: 7.1164%
sensor_event_duration Action=REMOVE_RESOURCE Type=Deployment (old: (186824.44/5000) 37.36, new (205114.76/5000) 41.02): change: 9.7901%
sensor_event_duration Action=REMOVE_RESOURCE Type=Pod (old: (536583.62/137324) 3.91, new (601624.97/125150) 4.81): change: 23.0280%
sensor_event_duration Action=UPDATE_RESOURCE Type=AlertResults (old: (9511.53/852) 11.16, new (7881.29/646) 12.20): change: 9.2834%
sensor_event_duration Action=UPDATE_RESOURCE Type=Binding (old: (6735.23/1728) 3.90, new (5309.38/1373) 3.87): change: -0.7879%
sensor_event_duration Action=UPDATE_RESOURCE Type=Deployment (old: (201249.87/27493) 7.32, new (330049.69/25075) 13.16): change: 79.8146%
sensor_event_duration Action=UPDATE_RESOURCE Type=Namespace (old: (159.48/30) 5.32, new (143.73/22) 6.53): change: 22.8946%
sensor_event_duration Action=UPDATE_RESOURCE Type=Pod (old: (2923.81/1294) 2.26, new (3133.88/1203) 2.61): change: 15.2928%
```

File: tools/generate-helpers/pg-table-bindings/schema.go.tpl [hunk 1]
```diff
@@ -75,18 +75,32 @@ var (
...
     type {{$schema.Table|upperCamelCase}} struct {
     {{- range $idx, $field := $schema.DBColumnFields }}
-        {{$field.ColumnName|upperCamelCase}} {{$field.ModelType}} `gorm:"column:{{$field.ColumnName|lowerCase}};type:{{$field.SQLType}}{{if $field.Options.Unique}};unique{{end}}{{if $field.Options.PrimaryKey}};primaryKey{{end}}{{if $field.Options.Index}};index:{{$schema.Table|lowerCamelCase|lowerCase}}_{{$field.ColumnName|lowerCase}},type:{{$field.Options.Index}}{{end}}"`
+        {{$field.ColumnName|upperCamelCase}} {{$field.ModelType}} `gorm:"{{- /**/ -}}
+        column:{{$field.ColumnName|lowerCase}};{{- /**/ -}}
+        type:{{$field.SQLType}}{{if $field.Options.Unique}};unique{{end}}{{if $field.Options.PrimaryKey}};primaryKey{{end}}{{- /**/ -}}
+        {{if $field.Options.Index}};{{- /**/ -}}
+            index:{{$schema.Table|lowerCamelCase|lowerCase}}_{{$field.ColumnName|lowerCase}},{{- /**/ -}}
+            type:{{$field.Options.Index}}{{- /**/ -}}
+        {{else if $field|isSacScoping }};{{- /**/ -}}
```
[the remark this answers is not in the document]
Comment (janisz) [reply tools/generate-helpers/pg-table-bindings/schema.go.tpl]: Fixed

**Review by rhybrillou — APPROVED** [review]

**Comment by connorgorman** [comment]
Postgres operation diff
```
postgres_op_duration Type=ActiveComponent Operation=UpdateMany (old: (2780.78/397) 7.00, new (2777.84/401) 6.93): change: -1.1020%
postgres_op_duration Type=Alert Operation=Upsert (old: (96545.06/42536) 2.27, new (131563.52/42321) 3.11): change: 36.9639%
postgres_op_duration Type=ClusterCVEEdges Operation=RemoveMany (old: (5.46/7) 0.78, new (3.23/7) 0.46): change: -40.8913%
postgres_op_duration Type=ClusterCVEs Operation=RemoveMany (old: (6.26/7) 0.89, new (4.24/7) 0.61): change: -32.2371%
postgres_op_duration Type=ClusterHealthStatus Operation=Upsert (old: (221.70/94) 2.36, new (210.41/95) 2.21): change: -6.0904%
postgres_op_duration Type=Deployment Operation=Remove (old: (11416.24/5033) 2.27, new (12281.96/5032) 2.44): change: 7.6046%
postgres_op_duration Type=Deployment Operation=Upsert (old: (363378.52/62530) 5.81, new (364389.97/61085) 5.97): change: 2.6505%
postgres_op_duration Type=Image Operation=Upsert (old: (4413.17/108) 40.86, new (4752.80/107) 44.42): change: 8.7025%
postgres_op_duration Type=IntegrationHealth Operation=Upsert (old: (767.60/190) 4.04, new (897.46/188) 4.77): change: 18.1621%
postgres_op_duration Type=K8SRole Operation=Remove (old: (239.90/133) 1.80, new (201.27/133) 1.51): change: -16.1020%
postgres_op_duration Type=K8SRole Operation=Upsert (old: (14617.69/5000) 2.92, new (12849.57/5000) 2.57): change: -12.0957%
postgres_op_duration Type=K8SRoleBinding Operation=Remove (old: (237.87/126) 1.89, new (203.88/126) 1.62): change: -14.2893%
postgres_op_duration Type=K8SRoleBinding Operation=Upsert (old: (23427.20/6728) 3.48, new (20219.24/6373) 3.17): change: -8.8857%
postgres_op_duration Type=NamespaceMetadata Operation=Remove (old: (9.80/5) 1.96, new (7.50/5) 1.50): change: -23.4612%
postgres_op_duration Type=NamespaceMetadata Operation=Upsert (old: (1332.06/44) 30.27, new (1302.01/34) 38.29): change: 26.4925%
postgres_op_duration Type=NetworkBaseline Operation=RemoveMany (old: (8208.85/5021) 1.63, new (8644.42/5021) 1.72): change: 5.3060%
postgres_op_duration Type=NetworkBaseline Operation=UpdateMany (old: (123212.15/9775) 12.60, new (129672.01/9718) 13.34): change: 5.8602%
postgres_op_duration Type=NetworkFlow Operation=RemoveFlowsByDeployment (old: (12723.32/5033) 2.53, new (14101.25/5032) 2.80): change: 10.8519%
postgres_op_duration Type=NetworkFlow Operation=UpdateMany (old: (4436.48/59) 75.19, new (4414.45/55) 80.26): change: 6.7399%
postgres_op_duration Type=NetworkPolicy Operation=Remove (old: (18.91/7) 2.70, new (9.73/7) 1.39): change: -48.5462%
postgres_op_duration Type=Node Operation=Upsert (old: (478168.69/1003) 476.74, new (397477.25/1003) 396.29): change: -16.8751%
postgres_op_duration Type=Pod Operation=Remove (old: (287985.17/137369) 2.10, new (319425.41/125196) 2.55): change: 21.7020%
postgres_op_duration Type=Pod Operation=Upsert (old: (328739.80/138618) 2.37, new (371970.01/126353) 2.94): change: 24.1337%
postgres_op_duration Type=PolicyCategory Operation=Upsert (old: (17.57/12) 1.46, new (16.59/12) 1.38): change: -5.5717%
postgres_op_duration Type=ProcessBaseline Operation=Remove (old: (24715.52/15000) 1.65, new (26412.06/15000) 1.76): change: 6.8643%
postgres_op_duration Type=ProcessBaseline Operation=Upsert (old: (25700.51/15115) 1.70, new (35704.67/15126) 2.36): change: 38.8249%
postgres_op_duration Type=ProcessBaselineResults Operation=Remove (old: (7469.80/5033) 1.48, new (8143.80/5032) 1.62): change: 9.0446%
postgres_op_duration Type=ProcessBaselineResults Operation=Upsert (old: (58507.66/21488) 2.72, new (70795.24/22700) 3.12): change: 14.5411%
postgres_op_duration Type=ProcessIndicator Operation=Remove (old: (240335.33/137369) 1.75, new (275068.35/125196) 2.20): change: 25.5802%
postgres_op_duration Type=ProcessIndicator Operation=RemoveMany (old: (2570.91/178) 14.44, new (2619.26/167) 15.68): change: 8.5915%
postgres_op_duration Type=ProcessIndicator Operation=UpdateMany (old: (13057.60/186) 70.20, new (13981.47/184) 75.99): change: 8.2392%
postgres_op_duration Type=Risk Operation=Remove (old: (8340.56/5033) 1.66, new (9084.16/5032) 1.81): change: 8.9371%
postgres_op_duration Type=Risk Operation=Upsert (old: (285786.92/106467) 2.68, new (289995.32/95061) 3.05): change: 13.6479%
postgres_op_duration Type=Secret Operation=Remove (old: (29.01/17) 1.71, new (39.16/17) 2.30): change: 34.9634%
postgres_op_duration Type=ServiceAccount Operation=Remove (old: (101.42/57) 1.78, new (211.68/57) 3.71): change: 108.7118%
postgres_op_duration Type=ServiceAccount Operation=Upsert (old: (14955.79/5000) 2.99, new (12731.25/5000) 2.55): change: -14.8741%
```

**Comment by connorgorman** [comment]
@janisz Do we have a comparison to the sheet that @rhybrillou created to show the speed ups?

**Comment by janisz** [comment]
I based on what @rhybrillou told me about his tests with indexes. 
@rhybrillou Do you have hard data? 😄

**Comment by rhybrillou** [comment]
I mixed all improvements in the last figures I collected.
The SAC index on deployments showed a cost about 50% lower on some of the previously slow queries when using the SAC index compared to when not.
I'll see what data I can extract next week.

**Pull request merged**